dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/ExecutionModel.java

```diff
@@ -361,7 +361,7 @@ private void addEvent(Event e, int globalId, int localId) {
 
         } else if (data.isFence()) {
             // ===== Fences =====
-            String name = ((Fence)data.getEvent()).getName();
+            String name = ((TaggedEvent)data.getEvent()).getName();
             fenceMap.computeIfAbsent(name, key -> new HashSet<>()).add(data);
```

Comment:
This code certainly violates the fact that `TaggedEvents` should be identified by their tags only (semantics-wise).
IIRC, this is done like this right now because we treat `fencerel(X)` as a base relation rather than a short-cut for `po;[X];po`.
I would put a `FIXME` there so we fix this eventually. You should double check what `data.isFence()` does and if it still applies for general `TaggedEvents`!

Reply:
I don't fully understand what your concern is.
```
public boolean isFence() { return event.hasTag(FENCE); }
```
thus the event is being identified by its tag (and if that invariant is satisfied, I don't see the problem with the cast)

Reply:
The problem is that the code assumes not only that the event has a fence tag but also that it is a `TaggedEvent`.
What if you had something like a `FencedLoad` that acts as a fence and a load simultaneously? In fact, acquire/release memory accesses fit into that picture (they just don't have the `F` tag).

Reply:
ok, now I got it. I agree, but the old code had the same issue. For now I will simply add a FIXME